File: ai-service/app/scheduling/slots.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from datetime import UTC, date, datetime, time, timedelta
from typing import Any


WEEKDAY_NAMES = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]


def resolve_scheduled_at(
    *,
    params: dict[str, Any],
    campaign_format: dict[str, Any] | None,
    context: dict[str, Any],
) -> datetime | None:
    preferred_time = preferred_posting_time(context, campaign_format)
    target_date = params.get("target_date")
    if isinstance(target_date, str) and target_date.strip():
        selected_date = date.fromisoformat(target_date.strip())
        return datetime.combine(selected_date, preferred_time, tzinfo=UTC)

    if campaign_format is None:
        return datetime.combine(_utcnow().date() + timedelta(days=1), preferred_time, tzinfo=UTC)

    weekday = campaign_format.get("weekday")
    if weekday is None:
        return datetime.combine(_utcnow().date() + timedelta(days=1), preferred_time, tzinfo=UTC)

    target_go_weekday = _coerce_int(weekday)
    if target_go_weekday is None:
        return None

    occupied_dates = occupied_campaign_dates(context, target_go_weekday)
    now = _utcnow().astimezone(UTC)
    for offset in range(0, 366):
        candidate_date = now.date() + timedelta(days=offset)
        if _go_weekday(datetime.combine(candidate_date, time.min, tzinfo=UTC)) != target_go_weekday:
            continue
        slot = datetime.combine(candidate_date, preferred_time, tzinfo=UTC)
        if slot <= now:
            continue
        if candidate_date.isoformat() in occupied_dates:
            continue
        return slot
    return None


def preferred_posting_time(context: dict[str, Any], campaign_format: dict[str, Any] | None) -> time:
    engagement_hours = context.get("engagement_hours") or context.get("engagementHours") or []
    peak_hour = _best_engagement_hour(engagement_hours)
    if peak_hour is not None:
        return time(hour=peak_hour, minute=0)

    scheduling = _team_scheduling_preferences(context)
    weekday = _coerce_int((campaign_format or {}).get("weekday"))
    posting_windows = scheduling.get("posting_windows") or []
    if weekday is not None:
        for window in posting_windows:
            if not isinstance(window, Mapping):
                continue
            if _coerce_int(window.get("weekday")) == weekday:
                parsed = _parse_clock(window.get("start"))
                if parsed is not None:
                    return parsed
    default_timeslots = scheduling.get("default_timeslots") or []
    for item in default_timeslots:
        parsed = _parse_clock(item)
        if parsed is not None:
            return parsed
    return time(hour=9, minute=0)


def occupied_campaign_dates(context: dict[str, Any], weekday: int) -> set[str]:
    occupied: set[str] = set()
    upcoming = context.get("upcoming_posts") or context.get("upcomingPosts") or []
    for item in upcoming:
        if not isinstance(item, Mapping):
            continue
        scheduled_at = item.get("scheduled_at") or item.get("scheduledAt")
        if not isinstance(scheduled_at, str) or not scheduled_at.strip():
            continue
        try:
            parsed = datetime.fromisoformat(scheduled_at.replace("Z", "+00:00")).astimezone(UTC)
        except ValueError:
            continue
        if _go_weekday(parsed) == weekday:
            occupied.add(parsed.date().isoformat())
    return occupied


def _best_engagement_hour(buckets: Sequence[Any]) -> int | None:
    best_hour: int | None = None
    best_score = -1.0
    for item in buckets:
        if not isinstance(item, Mapping):
            continue
        hour = _coerce_int(item.get("hour") or item.get("hour_utc") or item.get("hourUTC"))
        score_raw = item.get("score")
        try:
            score = float(score_raw)
        except (TypeError, ValueError):
            score = 0.0
        if hour is None or hour < 0 or hour > 23:
            continue
        if score > best_score:
            best_score = score
            best_hour = hour
    return best_hour


def _team_scheduling_preferences(context: dict[str, Any]) -> dict[str, Any]:
    team = context.get("team") or {}
    if isinstance(team, Mapping):
        prefs = team.get("scheduling_preferences") or team.get("schedulingPreferences") or {}
        if isinstance(prefs, Mapping):
            return dict(prefs)
    return {}


def _parse_clock(raw_value: Any) -> time | None:
    if not isinstance(raw_value, str) or not raw_value.strip():
        return None
    parts = raw_value.strip().split(":")
    if len(parts) < 2:
        return None
    try:
        hour = int(parts[0])
        minute = int(parts[1])
    except ValueError:
        return None
    if hour < 0 or hour > 23 or minute < 0 or minute > 59:
        return None
    return time(hour=hour, minute=minute)


def _coerce_int(value: Any) -> int | None:
    if value is None:
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _go_weekday(value: datetime) -> int:
    return (value.weekday() + 1) % 7


def _utcnow() -> datetime:
    return datetime.now(UTC)


def format_datetime(value: datetime | None) -> str | None:
    if value is None:
        return None
    return value.astimezone(UTC).isoformat().replace("+00:00", "Z")

File: internal/ai/slots.go
package ai

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"git.f4mily.net/goloom/internal/domain"
)

var weekdayNamesEN = []string{"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"}
var weekdayNamesDE = []string{"Sonntag", "Montag", "Dienstag", "Mittwoch", "Donnerstag", "Freitag", "Samstag"}

// nowFunc is swapped in tests.
var nowFunc = func() time.Time { return time.Now().UTC() }

// formatScheduleLabel renders a human-readable schedule label for prompts (UTC).
func formatScheduleLabel(isoValue, language string) string {
	raw := strings.TrimSpace(isoValue)
	if raw == "" {
		return ""
	}
	dt, err := parseISOTime(raw)
	if err != nil {
		return raw
	}
	dt = dt.UTC()
	weekdayIndex := int(dt.Weekday())
	if strings.HasPrefix(strings.ToLower(strings.TrimSpace(language)), "de") {
		return fmt.Sprintf("%s, %02d.%02d.%d, %02d:%02d UTC", weekdayNamesDE[weekdayIndex], dt.Day(), int(dt.Month()), dt.Year(), dt.Hour(), dt.Minute())
	}
	return fmt.Sprintf("%s, %s, %s UTC", weekdayNamesEN[weekdayIndex], dt.Format("2006-01-02"), dt.Format("15:04"))
}

func parseISOTime(raw string) (time.Time, error) {
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339, "2006-01-02T15:04:05", "2006-01-02 15:04:05"} {
		if t, err := time.Parse(layout, raw); err == nil {
			return t, nil
		}
	}
	return time.Time{}, fmt.Errorf("unparseable time %q", raw)
}

func formatDatetime(value *time.Time) string {
	if value == nil || value.IsZero() {
		return ""
	}
	return value.UTC().Format("2006-01-02T15:04:05Z07:00")
}

// resolveScheduledAt picks a publication slot for generated posts.
func resolveScheduledAt(params params, campaignFormat *domain.CampaignFormat, context domain.AIContext) *time.Time {
	preferred := preferredPostingTime(context, campaignFormat)
	if targetDate := params.str("target_date"); targetDate != "" {
		if day, err := time.Parse("2006-01-02", strings.TrimSpace(targetDate)); err == nil {
			at := time.Date(day.Year(), day.Month(), day.Day(), preferred.hour, preferred.minute, 0, 0, time.UTC)
			return &at
		}
	}

	now := nowFunc()
	if campaignFormat == nil || campaignFormat.Weekday == nil {
		day := now.AddDate(0, 0, 1)
		at := time.Date(day.Year(), day.Month(), day.Day(), preferred.hour, preferred.minute, 0, 0, time.UTC)
		return &at
	}

	targetWeekday := *campaignFormat.Weekday
	occupied := occupiedCampaignDates(context, targetWeekday)
	for offset := 0; offset < 366; offset++ {
		candidate := now.AddDate(0, 0, offset)
		if int(candidate.Weekday()) != targetWeekday {
			continue
		}
		slot := time.Date(candidate.Year(), candidate.Month(), candidate.Day(), preferred.hour, preferred.minute, 0, 0, time.UTC)
		if !slot.After(now) {
			continue
		}
		if occupied[slot.Format("2006-01-02")] {
			continue
		}
		return &slot
	}
	return nil
}

// NextCampaignSlot returns the next free publication slot for a campaign
// format, honoring its weekday, the team's scheduling preferences, and
// already scheduled posts.
func NextCampaignSlot(context domain.AIContext, format *domain.CampaignFormat) *time.Time {
	return resolveScheduledAt(params{}, format, context)
}

type clockTime struct {
	hour   int
	minute int
}

func preferredPostingTime(context domain.AIContext, campaignFormat *domain.CampaignFormat) clockTime {
	if hour, ok := bestEngagementHour(context.EngagementHours); ok {
		return clockTime{hour: hour}
	}

	prefs := context.Team.SchedulingPrefs
	if campaignFormat != nil && campaignFormat.Weekday != nil {
		for _, window := range prefs.PostingWindows {
			if window.Weekday == *campaignFormat.Weekday {
				if parsed, ok := parseClock(window.Start); ok {
					return parsed
				}
			}
		}
	}
	for _, slot := range prefs.DefaultTimeslots {
		if parsed, ok := parseClock(slot); ok {
			return parsed
		}
	}
	return clockTime{hour: 9}
}

func occupiedCampaignDates(context domain.AIContext, weekday int) map[string]bool {
	occupied := map[string]bool{}
	for _, post := range context.UpcomingPosts {
		at := post.ScheduledAt.UTC()
		if at.IsZero() {
			continue
		}
		if int(at.Weekday()) == weekday {
			occupied[at.Format("2006-01-02")] = true
		}
	}
	return occupied
}

func bestEngagementHour(buckets []domain.EngagementHourBucket) (int, bool) {
	bestHour := -1
	bestScore := int64(-1)
	for _, bucket := range buckets {
		if bucket.HourUTC < 0 || bucket.HourUTC > 23 {
			continue
		}
		if bucket.Score > bestScore {
			bestScore = bucket.Score
			bestHour = bucket.HourUTC
		}
	}
	return bestHour, bestHour >= 0
}

func parseClock(raw string) (clockTime, bool) {
	parts := strings.Split(strings.TrimSpace(raw), ":")
	if len(parts) < 2 {
		return clockTime{}, false
	}
	hour, err1 := strconv.Atoi(parts[0])
	minute, err2 := strconv.Atoi(parts[1])
	if err1 != nil || err2 != nil || hour < 0 || hour > 23 || minute < 0 || minute > 59 {
		return clockTime{}, false
	}
	return clockTime{hour: hour, minute: minute}, true
}
